base_layer/blockchain/src/blockchainstate.rs

```diff
@@ -22,45 +22,169 @@
+    pub fn change_pruning_horizon(&mut self, new_pruning_horizon: usize) {
+        self.outputs.set_pruning_horizon(new_pruning_horizon);
     }
 
     /// This function  will process a new block.
-    /// Note the block must have been validated by the chainstate before.
-    pub fn process_new_block(&self, new_block: &Block) -> Result<(), StateError> {
-        self.validate_new_block(new_block)
+    /// Note the block is consumed by the function.
+    pub fn process_new_block(&mut self, mut new_block: Block) -> Result<(), StateError> {
+        self.validate_new_block(&new_block)?;
+        self.prune_all_inputs(&new_block)?;
+        // All seems valid, lets add the objects to the state
+        let mut drainer = new_block.body.outputs.drain(..);
+        while let Some(output) = drainer.next() {
+            self.range_proofs.push(output.proof.clone())?;
+            self.outputs.push(output.into())?;
+        }
+        self.kernels.append(new_block.body.kernels)?;
+        self.headers.push(new_block.header)?;
+        // lets check states
+        self.check_mmr_states()?;
+        self.check_point_state()
     }
 
     /// This function will validate the block in terms of the current state.
-    pub fn validate_new_block(&self, _new_block: &Block) -> Result<(), StateError> {
+    pub fn validate_new_block(&self, new_block: &Block) -> Result<(), StateError> {
+        // we assume that we have atleast in block in the headers mmr even if this is the genesis one
+        if self.headers.get_last_added_object().unwrap().hash() != new_block.header.prev_hash {
+            return Err(StateError::OrphanBlock);
+        }
+        new_block
+            .check_internal_consistency()
+            .map_err(StateError::InvalidBlock)?;
+        Ok(())
+    }
+
+    /// On validation error the chain state should be reset to the last validly saved state
+    pub fn reset_chain_state(&mut self) -> Result<(), StateError> {
+        self.headers.ff_to_head(&mut self.store)?;
+        self.range_proofs.ff_to_head(&mut self.store)?;
+        self.outputs.ff_to_head(&mut self.store)?;
+        self.kernels.ff_to_head(&mut self.store)?;
+        Ok(())
+    }
+
+    /// This function test if all the inputs are valid unpruned out puts and flags them as pruned.
+    /// This will return an error if any of the inputs where pruned or unknown
+    fn prune_all_inputs(&mut self, new_block: &Block) -> Result<(), StateError> {
+        for input in &new_block.body.inputs {
+            let hash = input.hash();
+            self.outputs
+                .prune_object_hash(&hash)
+                .map_err(StateError::SpentUnknownCommitment)?;
+        }
+        Ok(())
+    }
+
+    /// This function is just a wrapper function to call checkpoint on all the MMR's
+    fn check_point_state(&mut self) -> Result<(), StateError> {
+        self.headers.checkpoint()?;
+        self.kernels.checkpoint()?;
+        self.outputs.checkpoint()?;
```

Comment:
To help with disambiguation, we should perhaps call this `utxos` or `unspent_outputs`; right now transaction outputs are also just called `outputs`